src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.netcoreapp.cs

```diff
@@ -24,5 +36,166 @@ public static void Create(this DirectoryInfo directoryInfo, DirectorySecurity di
+        /// <exception cref="ArgumentException">Invalid handle.
+        ///-or-
+        /// Handle does not support asynchronous operations. The parameters to the <see cref="FileStream" /> constructor may need to be changed to indicate that the handle was opened synchronously (that is, it was not opened for overlapped I/O).
+        /// </exception>
+        /// <exception cref="ObjectDisposedException">Cannot access a closed file.</exception>
+        public static FileStream Create(this FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options, FileSecurity fileSecurity)
+        {
+            if (fileInfo == null)
+            {
+                throw new ArgumentNullException(nameof(fileInfo));
+            }
+
+            if (fileSecurity == null)
+            {
+                throw new ArgumentNullException(nameof(fileSecurity));
+            }
+
+            // don't include inheritable in our bounds check for share
+            FileShare tempshare = share & ~FileShare.Inheritable;
+
+            string badArg = null;
+            if (mode < FileMode.CreateNew || mode > FileMode.Append)
+            {
+                badArg = nameof(mode);
+            }
+            else if (tempshare < FileShare.None || tempshare > (FileShare.ReadWrite | FileShare.Delete))
+            {
+                badArg = nameof(share);
+            }
+            if (badArg != null)
+            {
+                throw new ArgumentOutOfRangeException(badArg, SR.ArgumentOutOfRange_Enum);
+            }
+
+            // We need to check only the supported attributes
+            // NOTE: any change to FileOptions enum needs to be matched here in the error validation
+            if (options != FileOptions.None &&
+                (options & ~(FileOptions.WriteThrough | FileOptions.Asynchronous | FileOptions.RandomAccess | FileOptions.DeleteOnClose | FileOptions.SequentialScan | FileOptions.Encrypted | (FileOptions)0x20000000 /* NoBuffering */)) != 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(options), SR.ArgumentOutOfRange_Enum);
+            }
+
+            if (bufferSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(bufferSize), SR.ArgumentOutOfRange_NeedPosNum);
+            }
+
+            SafeFileHandle handle = null;
+            try
+            {
+                handle = CreateFileOpenHandle(fileInfo.FullName, mode, rights, share, options, fileSecurity);
+                return new FileStream(handle, GetFileStreamFileAccess(rights), bufferSize, (options & FileOptions.Asynchronous) != 0);
+            }
+            catch
+            {
+                // If anything goes wrong while setting up the stream, make sure we deterministically dispose of the opened handle.
+                if (handle != null)
+                {
+                    handle.Dispose();
+                    handle = null;
```

Comment:
@JeremyKuhne is `handle = null` really necessary?

Reply:
Not now that it's a local.

Reply:
Or just move the `CreateFileOpenHandle` out of the try. Now you don't need to check it for null (as long as you trust that nobody changes CreateFileOpenHandle such that it can somehow return null)

Reply:
Thanks, I made the suggested changes.
Eventually we will add nullability to this, and I suppose that if someone decides to modify the method to potentially return a null, it would fail.